File: api-java/src/test/java/com/github/pflooky/datacaterer/api/java/PlanRunTest.java
package com.github.pflooky.datacaterer.api.java;

import com.github.pflooky.datacaterer.api.java.model.FieldBuilder;
import com.github.pflooky.datacaterer.api.java.model.config.connection.ConnectionTaskBuilder;
import com.github.pflooky.datacaterer.api.model.Constants;
import org.junit.Test;

import java.util.Objects;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class PlanRunTest {

    private class BasePlanRun extends PlanRun {}

    @Test
    public void canCreatePlanRunWithSecondTaskInheritConfigFromFirstTask() {
        var baseResult = new BasePlanRun();
        ConnectionTaskBuilder myPostgres = baseResult.postgres("my_postgres", "jdbc:url")
                .table("account.accounts")
                .schema(new FieldBuilder().name("account_id"));
        ConnectionTaskBuilder mySecondPostgres = baseResult.postgres(myPostgres)
                .table("account.transactions")
                .schema(new FieldBuilder().name("txn_id"));

        baseResult.execute(myPostgres, mySecondPostgres);
        var result = baseResult.getPlan();

        assertEquals(2, result._tasks().size());
        var firstStep = result._tasks().head().steps().head();
        var secondStep = result._tasks().last().steps().head();
        var accountStep = firstStep.options().get(Constants.JDBC_TABLE()).contains("account.accounts") ? firstStep : secondStep;
        var txnStep  = firstStep.options().get(Constants.JDBC_TABLE()).contains("account.transactions") ? firstStep : secondStep;
        var accountSchema = accountStep.schema();
        var txnSchema = txnStep.schema();
        assertTrue(accountSchema.fields().isDefined());
        assertEquals(1, accountSchema.fields().get().size());
        assertTrue(accountSchema.fields().get().exists(f -> Objects.equals(f.name(), "account_id")));
        assertTrue(txnSchema.fields().isDefined());
        assertEquals(1, txnSchema.fields().get().size());
        assertTrue(txnSchema.fields().get().exists(f -> Objects.equals(f.name(), "txn_id")));

        assertEquals(1, result._configuration().connectionConfigByName().size());
        assertTrue(result._configuration().connectionConfigByName().contains("my_postgres"));
        assertTrue(result._configuration().connectionConfigByName().get("my_postgres").get().contains(Constants.URL()));
        assertTrue(result._configuration().connectionConfigByName().get("my_postgres").get().get(Constants.URL()).contains("jdbc:url"));
        assertTrue(result._validations().isEmpty());
    }

}

File: api/src/test/scala/com/github/pflooky/datacaterer/api/PlanRunTest.scala
package com.github.pflooky.datacaterer.api

import com.github.pflooky.datacaterer.api.model.Constants.{CSV, FORMAT, JDBC_TABLE, PATH, URL}
import com.github.pflooky.datacaterer.api.model.ExpressionValidation
import org.junit.runner.RunWith
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class PlanRunTest extends AnyFunSuite {

  test("Can create plan with each type of connection") {
    val result = new PlanRun {
      val mySchema = schema.addFields(field.name("account_id"))
      val myCsv = csv("my_csv", "/my/csv").schema(mySchema)
      val myJson = json("my_json", "/my/json").schema(mySchema)
      val myParquet = parquet("my_parquet", "/my/parquet").schema(mySchema)
      val myOrc = orc("my_orc", "/my/orc").schema(mySchema)
      val myPostgres = postgres("my_postgres").table("account").schema(mySchema)
      val myMySql = mysql("my_mysql").table("transaction").schema(mySchema)
      val myCassandra = cassandra("my_cassandra").table("account", "accounts").schema(mySchema)
      val mySolace = solace("my_solace").destination("solace_topic").schema(mySchema)
      val myKafka = kafka("my_kafka").topic("kafka_topic").schema(mySchema)
      val myHttp = http("my_http").schema(mySchema)

      execute(myCsv, myJson, myParquet, myOrc, myPostgres, myMySql, myCassandra, mySolace, myKafka, myHttp)
    }

    val dsNames = List("my_csv", "my_json", "my_parquet", "my_orc", "my_postgres", "my_mysql", "my_cassandra", "my_solace", "my_kafka", "my_http")
    assert(result._plan.tasks.size == 10)
    assert(result._plan.tasks.map(_.dataSourceName) == dsNames)
    assert(result._configuration.connectionConfigByName.size == 10)
    assert(result._configuration.connectionConfigByName.keys.forall(dsNames.contains))
    assert(result._tasks.size == 10)
  }

  test("Can create plan using same connection details from another step") {
    val result = new PlanRun {
      val myPostgresAccount = postgres("my_postgres", "my_postgres_url")
        .table("account.accounts")
        .schema(field.name("account_id"))
      val myPostgresTransaction = postgres(myPostgresAccount)
        .table("account.transactions")
        .schema(field.name("txn_id"))

      execute(myPostgresAccount, myPostgresTransaction)
    }

    assert(result._plan.tasks.size == 2)
    assert(result._plan.tasks.map(_.dataSourceName).forall(_ == "my_postgres"))
    assert(result._configuration.connectionConfigByName.size == 1)
    assert(result._configuration.connectionConfigByName.contains("my_postgres"))
    assert(result._configuration.connectionConfigByName("my_postgres").contains(URL))
    assert(result._configuration.connectionConfigByName("my_postgres").get(URL).contains("my_postgres_url"))
    assert(result._tasks.size == 2)
    val steps = result._tasks.flatMap(_.steps)
    val resAccount = steps.filter(s => s.options.get(JDBC_TABLE).contains("account.accounts")).head
    assert(resAccount.schema.fields.isDefined)
    assert(resAccount.schema.fields.get.size == 1)
    assert(resAccount.schema.fields.get.head.name == "account_id")
    val resTxn = steps.filter(s => s.options.get(JDBC_TABLE).contains("account.transactions")).head
    assert(resTxn.schema.fields.isDefined)
    assert(resTxn.schema.fields.get.size == 1)
    assert(resTxn.schema.fields.get.head.name == "txn_id")
    assert(result._validations.isEmpty)
  }

  test("Can create plan with validations for one data source") {
    val result = new PlanRun {
      val myCsv = csv("my_csv", "/my/data/path")
        .schema(field.name("account_id"))
        .validations(validation.expr("account_id != ''"))

      execute(myCsv)
    }

    assert(result._validations.size == 1)
    assert(result._validations.head.dataSources.size == 1)
    val dsValidation = result._validations.head.dataSources.head
    assert(dsValidation._1 == "my_csv")
    assert(dsValidation._2.options.size == 2)
    assert(dsValidation._2.options.get(PATH).contains("/my/data/path"))
    assert(dsValidation._2.options.get(FORMAT).contains("csv"))
    assert(dsValidation._2.validations.size == 1)
    assert(dsValidation._2.validations.head.validation.isInstanceOf[ExpressionValidation])
    val expressionValidation = dsValidation._2.validations.head.validation.asInstanceOf[ExpressionValidation]
    assert(expressionValidation.expr == "account_id != ''")
  }

  test("Can create plan with multiple validations for one data source") {
    val result = new PlanRun {
      val myPostgresAccount = postgres("my_postgres")
        .table("account.accounts")
        .validations(validation.expr("account_id != ''"))
      val myPostgresTransaction = postgres("my_postgres")
        .table("account", "transactions")
        .validations(validation.expr("txn_id IS NOT NULL"))

      execute(myPostgresAccount, myPostgresTransaction)
    }

    assert(result._validations.size == 1)
    assert(result._validations.head.dataSources.size == 1)
    val dsValidation = result._validations.head.dataSources.head
    assert(dsValidation._1 == "my_postgres")
    assert(dsValidation._2.options.nonEmpty)
    assert(dsValidation._2.options.get(FORMAT).contains("jdbc"))
    assert(dsValidation._2.validations.size == 2)
    assert(dsValidation._2.validations.exists(v => v.validation.asInstanceOf[ExpressionValidation].expr == "account_id != ''"))
    assert(dsValidation._2.validations.exists(v => v.validation.asInstanceOf[ExpressionValidation].expr == "txn_id IS NOT NULL"))
  }

  test("Can create plan with validations only defined") {
    val result = new PlanRun {
      val myCsv = csv("my_csv", "/my/csv")
        .validations(validation.expr("account_id != 'acc123'"))

      execute(myCsv)
    }

    assert(result._tasks.size == 1)
    assert(result._validations.size == 1)
    assert(result._validations.head.dataSources.contains("my_csv"))
    val validRes = result._validations.head.dataSources("my_csv")
    assert(validRes.validations.size == 1)
    assert(validRes.validations.head.validation.asInstanceOf[ExpressionValidation].expr == "account_id != 'acc123'")
    assert(validRes.options.size == 2)
    assert(validRes.options(FORMAT) == CSV)
    assert(validRes.options(PATH) == "/my/csv")
  }


}